File: go/structures/binary_heap.go
package structures

import (
	"errors"
	"strconv"
)

const (
	// MinHeap type.
	MinHeap = iota
	// MaxHeap type.
	MaxHeap = iota
)

// BinaryHeap can represent either a max or a min heap.
type BinaryHeap struct {
	HeapType uint32
	root     *BinaryTreeNode
	size     int
}

// Insert adds a new element to the heap.
func (heap *BinaryHeap) Insert(elem int) error {
	toInsert := &BinaryTreeNode{value: elem}
	if heap.root == nil {
		heap.root = toInsert
		heap.size = 1
		return nil
	}
	if heap.size == 1 && heap.getPriority(heap.root, toInsert) == 0 {
		return errors.New("Value already exists in heap: " + strconv.Itoa(elem))
	}
	insertAtNode := heap.getInsertNode()

	toInsert.parent = insertAtNode
	if isLeaf(insertAtNode) {
		insertAtNode.left = toInsert
	} else {
		insertAtNode.right = toInsert
	}
	for toInsert.parent != nil && heap.getPriority(toInsert.parent, toInsert) > 0 {
		if heap.getPriority(toInsert.parent, toInsert) == 0 {
			return errors.New("Value already exists in heap: " + strconv.Itoa(elem))
		}
		heap.swap(toInsert.parent, toInsert)
	}
	heap.size++
	return nil
}

// InsertAll inserts all elements in a slice sequentially.
func (heap *BinaryHeap) InsertAll(elems []int) error {
	var e error
	for _, elem := range elems {
		err := heap.Insert(elem)
		if err != nil {
			e = err
		}
	}
	return e
}

// RemoveTop removes and returns the root of the heap.
// The only error that can be thrown at this point is empty heap.
func (heap *BinaryHeap) RemoveTop() (int, error) {
	if heap.root == nil {
		return 0, errors.New("Heap is empty")
	}
	newRoot := heap.getRemoveNode()
	value := heap.root.value
	if isRoot(newRoot) {
		heap.Clear()
		return value, nil
	}
	if getRelationship(newRoot.parent, newRoot) == LeftChild {
		newRoot.parent.left = nil
	} else {
		newRoot.parent.right = nil
	}
	newRoot.parent = nil
	newRoot.left = heap.root.left
	newRoot.right = heap.root.right

	currentNode := newRoot
	didSwap := false
	for !isLeaf(currentNode) {
		// Because of the insertion strategy,
		// it is impossible for the left child to not exist if currentNode is not a leaf.
		var leftPriority, rightPriority int
		leftPriority = heap.getPriority(currentNode.left, currentNode)
		if currentNode.right != nil {
			rightPriority = heap.getPriority(currentNode.right, currentNode)
		}
		// Check for empty variable using 0 is safe because it same values would be caught in Insert.
		if rightPriority == 0 { // There is only one child (left child).
			if leftPriority < 0 {
				heap.swap(currentNode, currentNode.left)
				didSwap = true
			} else {
				break
			}
		} else { // Both children exist.
			if heap.getPriority(currentNode, currentNode.left) < 0 && heap.getPriority(currentNode, currentNode.right) < 0 {
				// Node is in the correct position.
				break
			}
			// Otherwise, keep bubbling.
			leftRightPriority := heap.getPriority(currentNode.left, currentNode.right)
			if leftRightPriority < 0 { // Left is smaller, so bubble left.
				heap.swap(currentNode, currentNode.left)
			} else {
				heap.swap(currentNode, currentNode.right)
			}
			didSwap = true
		}
	}
	// New root was immediately placed in the right position, swap pointers manually.
	// This can only occur when deleting from a heap of size 3 (resulting in size 2).
	if !didSwap {
		heap.root = newRoot
		heap.root.parent = nil
		// No need to check the right child, since it cannot exist in a heap of size 2.
		if heap.root.left != nil {
			heap.root.left.parent = heap.root
		}
	}
	heap.size--
	return value, nil
}

// Top returns the value at the top of the heap.
func (heap *BinaryHeap) Top() (int, error) {
	if heap.root == nil {
		return 0, errors.New("Heap is empty")
	}
	return heap.root.value, nil
}

// Clear removes all elements in the heap.
func (heap *BinaryHeap) Clear() {
	heap.root = nil
	heap.size = 0
}

// Size returns the number of elements in the heap.
func (heap *BinaryHeap) Size() int {
	return heap.size
}

// IsEmpty returns true if the heap has no elements.
func (heap *BinaryHeap) IsEmpty() bool {
	return heap.size == 0
}

func (heap *BinaryHeap) getInsertNode() *BinaryTreeNode {
	return getInsertNodeHelper(heap.root)
}

func getInsertNodeHelper(node *BinaryTreeNode) *BinaryTreeNode {
	if node.left == nil || node.right == nil {
		return node
	}
	if subtreeSize(node.left) <= subtreeSize(node.right) {
		return getInsertNodeHelper(node.left)
	}
	return getInsertNodeHelper(node.right)
}

func (heap *BinaryHeap) getRemoveNode() *BinaryTreeNode {
	return getRemoveNodeHelper(heap.root)
}

func getRemoveNodeHelper(node *BinaryTreeNode) *BinaryTreeNode {
	if isLeaf(node) {
		return node
	}
	if subtreeSize(node.left) <= subtreeSize(node.right) {
		return getRemoveNodeHelper(node.right)
	}
	return getRemoveNodeHelper(node.left)
}

func (heap *BinaryHeap) swap(parent *BinaryTreeNode, child *BinaryTreeNode) {
	var top, bottomLeft, bottomRight, sibling *BinaryTreeNode
	top = parent.parent
	bottomLeft = child.left
	bottomRight = child.right
	if getRelationship(parent, child) == LeftChild {
		sibling = parent.right
	} else {
		sibling = parent.left
	}

	child.parent = top
	if getRelationship(parent, child) == LeftChild {
		child.left = parent
	} else {
		child.right = parent
	}
	if top != nil {
		if getRelationship(top, parent) == LeftChild {
			top.left = child
		} else {
			top.right = child
		}
	}
	if sibling == nil {
		// If the sibling doesn't exist, then the child must be a left child.
		child.right = nil
	} else {
		if getRelationship(parent, sibling) == LeftChild {
			child.left = sibling
		} else {
			child.right = sibling
		}
		sibling.parent = child
	}
	parent.left = bottomLeft
	if bottomLeft != nil {
		bottomLeft.parent = parent
	}
	parent.right = bottomRight
	if bottomRight != nil {
		bottomRight.parent = parent
	}

	parent.parent = child
	if isRoot(child) {
		heap.root = child
	}
}

// Lowest priority goes to the top of the heap.
func (heap *BinaryHeap) getPriority(a *BinaryTreeNode, b *BinaryTreeNode) int {
	if heap.HeapType == MinHeap {
		return a.value - b.value
	}
	return b.value - a.value
}

File: ruby/src/data-structures/binary_heap.rb
# frozen-string-literal: true

require 'tree'

EMPTY_HEAP_EXCEPTION = 'Heap is empty'

module DataStructures
  module BinaryHeapType
    MIN = 'min'
    MAX = 'max'
  end

  class BinaryHeap
    attr_reader :size

    def initialize(heap_type, predicate)
      @root = nil
      @size = 0
      @heap_type = heap_type
      @predicate = predicate
    end

    def insert(elem)
      to_insert = BinaryTreeNode.new(elem, nil, nil, nil)
      if @root.nil?
        @root = to_insert
        @size += 1
        return true
      end

      insert_at = insert_node
      to_insert.parent = insert_at
      if leaf?(insert_at)
        insert_at.left = to_insert
      else
        insert_at.right = to_insert
      end

      swap(to_insert.parent, to_insert) while !to_insert.parent.nil? && priority(to_insert.parent, to_insert).positive?
      @size += 1
      true
    end

    def insert_all(elems)
      result = true
      elems.each do |elem|
        success = insert(elem)
        result &&= success
      end
      result
    end

    # Removes and returns the root of the heap.
    def remove_top
      return raise EMPTY_HEAP_EXCEPTION if @root.nil?

      new_root = remove_node
      value = @root.value
      if root?(new_root)
        @root = nil
        @size = 0
        return value
      end
      if relationship(new_root.parent, new_root) == BinaryNodeRelationship::LEFT_CHILD
        new_root.parent.left = nil
      else
        new_root.parent.right = nil
      end
      new_root.parent = nil
      new_root.left = @root.left
      new_root.right = @root.right

      current = new_root
      did_swap = false
      until leaf?(current)
        # Because of the insertion strategy,
        # it is impossible for the left child to not exist if the current is not a leaf.
        left_priority = priority(current.left, current)
        right_priority = current.right.nil? ? nil : priority(current.right, current)

        if right_priority.nil? # There is only a left child.
          break unless left_priority.negative?

          swap(current, current.left)
        else # Both children exist.
          # Node is in the correct position.
          break if priority(current, current.left).negative? && priority(current, current.right).negative?

          # Otherwise, keep bubbling.
          left_right_priority = priority(current.left, current.right)
          # Bubble towards the smaller priority.
          swap(current, left_right_priority.negative? ? current.left : current.right)
        end
        did_swap = true
      end
      # New root was immediately placed in the correct position, swap pointers manually.
      # This can only occur when deleting from a heap of size 3 (resulting in size 2).
      unless did_swap
        @root = new_root
        @root.parent = nil
        # No need to check the right child, since it cannot exist in a heap of size 3.
        @root.left.parent = @root unless @root.left.nil?
      end
      @size -= 1
      value
    end

    def top
      return raise EMPTY_HEAP_EXCEPTION if @root.nil?

      @root.value
    end

    def clear
      @root = nil
      @size = 0
    end

    def empty?
      @size.zero?
    end

    private

    # Swap a child with its parent.
    def swap(parent, child)
      top = parent.parent
      sibling = relationship(parent, child) == BinaryNodeRelationship::LEFT_CHILD ? parent.right : parent.left
      bottom_left = child.left
      bottom_right = child.right

      child.parent = top
      if relationship(parent, child) == BinaryNodeRelationship::LEFT_CHILD
        child.left = parent
      else
        child.right = parent
      end

      unless top.nil?
        if relationship(top, parent) == BinaryNodeRelationship::LEFT_CHILD
          top.left = child
        else
          top.right = child
        end
      end

      if sibling.nil?
        # If the sibling node doesn't exist, then the child must be a left child.
        child.right = nil
      else
        if relationship(parent, sibling) == BinaryNodeRelationship::LEFT_CHILD
          child.left = sibling
        else
          child.right = sibling
        end
        sibling.parent = child
      end
      parent.left = bottom_left
      bottom_left.parent = parent unless bottom_left.nil?
      parent.right = bottom_right
      bottom_right.parent = parent unless bottom_right.nil?

      parent.parent = child
      @root = child if root?(child)
    end

    # Counts the number of nodes in the subtree rooted at a given node.
    def subtree_size(node)
      return 0 if node.nil?

      1 + subtree_size(node.left) + subtree_size(node.right)
    end

    # Returns the node where the insert should happen.
    def insert_node
      insert_node_helper(@root)
    end

    def insert_node_helper(node)
      return node if node.left.nil? || node.right.nil?

      return insert_node_helper(node.left) if subtree_size(node.left) <= subtree_size(node.right)

      insert_node_helper(node.right)
    end

    # Returns the node that should replace the root following a call to remove_top.
    def remove_node
      remove_node_helper(@root)
    end

    def remove_node_helper(node)
      return node if leaf?(node)

      return remove_node_helper(node.right) if subtree_size(node.left) <= subtree_size(node.right)

      remove_node_helper(node.left)
    end

    # Completes the sentence: b is a's ______.
    def relationship(a, b)
      return BinaryNodeRelationship::LEFT_CHILD if !a.left.nil? && a.left == b

      return BinaryNodeRelationship::RIGHT_CHILD if !a.right.nil? && a.right == b

      return BinaryNodeRelationship::PARENT if !a.parent.nil? && a.parent == b

      BinaryNodeRelationship::NONE
    end

    def root?(node)
      node.parent.nil?
    end

    def leaf?(node)
      number_of_children(node).zero?
    end

    def number_of_children(node)
      (!node.left.nil? && 1 || 0) + (!node.right.nil? && 1 || 0)
    end

    # Lowest priority at the top.
    def priority(a, b)
      return raise "Element already exists in heap: #{a.value}" if @predicate.call(a.value, b.value).zero?

      return @predicate.call(a.value, b.value) if @heap_type == BinaryHeapType::MIN

      -@predicate.call(a.value, b.value)
    end
  end
end
